File: backend/orchestrator-prompt/system/crisis_management_scenarios.py
def get_crisis_management_scenarios_message(
    industry: str, businessSize: str, businessContent: str, selectedOptions: list[str], additionalContents: str, considerations: str
) -> str:
    """製造業向けの危機管理シナリオを作成する"""
    if additionalContents:
        output_format = (
            "### 出力形式: 危機管理シナリオとして、以下の項目を含めて出力してください。\n"
            "- リスクカテゴリ\n"
            f"- リスク内容: {additionalContents}\n"
            "- 想定ケース\n"
            "- 予防策\n"
        )
    else:
        output_format = (
            "### 出力形式: 危機管理シナリオとして、以下の項目を含めて出力してください。\n"
            "- リスクカテゴリ\n"
            "- リスク内容\n"
            "- 想定ケース\n"
            "- 予防策\n"
        )

    system_message = {
        "role": "system",
        "content": (
            "### 役割\n"
            "あなたは製造業のリスク対策に関する専門家です。\n\n"
            "### 目的\n"
            "未だ想定されていないリスクを特定し、それに対する危機管理シナリオを考案すること。\n\n"
            "### 指示\n"
            "以下の#流れに従って、詳細な危機管理シナリオを生成してください。\n"
            "ただし、余計な説明は不要です。\n"
            "回答は簡潔にまとめてください。各セクションは短く、要点を押さえた内容にしてください。\n"
            "### 流れ\n"
            "1. 危機管理の観点から、その業界と役割に特有の、まだ想定されていない「新たな事態」を1つ考え出してください。\n"
            "   この際、想定済みの事態と重複しないように注意してください。\n"
            "2. 1で挙げた「新たな事態」に対する予防策と、もし発生した場合の対応策を考えてください。\n"
            "3. 考えた「事態」と「対策」を基に、具体的な危機管理シナリオを作成してください。\n\n"
            f"{output_format}"
        ),
    }

    considerations_text = f"\n\n考慮事項：{considerations}" if considerations else ""

    user_message = {
        "role": "user",
        "content": (
            f"危機管理シナリオを作成してください：\n\n"
            f"業界・業種：{industry}\n"
            f"企業規模・拠点情報：{businessSize}\n"
            f"シナリオを作成する業務内容：{businessContent}\n"
            f"リスクカテゴリ：{selectedOptions}\n"
            f"リスク内容：{additionalContents}\n"
            f"考慮事項：{considerations_text}\n"
        ),
    }

    return [system_message, user_message]

File: frontend/src/app/(dashboard)/crisis-management-scenarios/_actions/crisis-management-scenarios.ts
'use server';

import { crisisManagementScenariosDB } from '@/app/_db/crisis-management-scenarios';
import { LLMserviceBackEndLog } from '@/app/_types/logs/llm-services';
import { getCurrentUser } from '@/app/_utils/auth';
import { getMessage } from '@/app/_utils/message';
import { useCaseAzureFunctions } from '../../../../../azure-functions';
import { crisisManagementScenariosContainer } from '../../../../../cosmos';

type IdeaResponse = {
  answer: string;
  log: LLMserviceBackEndLog<'idea'>;
};

type IdeaErrorResponse = {
  error: string;
};

export async function CrisisManagementScenarios(
  id: string,
  industry: string,
  businessSize: string,
  businessContent: string,
  selectedOptions: string[],
  additionalContents?: string,
  additionalConsiderations?: string
): Promise<IdeaResponse | IdeaErrorResponse> {
  const user = await getCurrentUser();
  try {
    const response = await useCaseAzureFunctions.sendJson<
      {
        industry: string;
        businessSize: string;
        businessContent: string;
        selectedOptions: string[];
        additionalContents?: string;
        additionalConsiderations?: string;
      },
      IdeaResponse
    >('crisis-management-scenarios', 'POST', {
      industry,
      businessSize,
      businessContent,
      selectedOptions,
      additionalContents,
      additionalConsiderations,
    });

    if (!response || !response.answer) {
      console.error('有効な回答がレスポンスに含まれていません:', response);
      return { error: getMessage('E_F_00110', '作成結果') };
    }

    // 応答を文字列として安全に処理
    const safeAnswer =
      typeof response.answer === 'string' ? response.answer : JSON.stringify(response.answer);

    // log
    try {
      await crisisManagementScenariosDB(crisisManagementScenariosContainer, {
        id,
        userId: user.id,
        useName: user.name,
        userEmail: user.email,
        userDepartmentName: user.departmentName,
        createdAt: new Date(),
        industry,
        businessSize,
        businessContent,
        selectedOptions,
        additionalContents,
        additionalConsiderations,
        outputForm: safeAnswer,
        log: response.log,
      });
    } catch (dbError) {
      console.error('データベース保存エラー:', dbError);
      // DBエラーでもフロントエンドには結果を返す
    }

    return {
      answer: safeAnswer,
      log: response.log,
    };
  } catch (error) {
    console.error('危機管理シナリオ作成error 詳細:', error);
    return { error: error instanceof Error ? error.message : getMessage('E_F_00110', '作成結果') };
  }
}

// ダミーデータ検証
// 'use server';

// import { crisisManagementScenariosDB } from '@/app/_db/crisis-management-scenarios';
// import { LLMserviceBackEndLog } from '@/app/_types/logs/llm-services';
// import { getCurrentUser } from '@/app/_utils/auth';
// import { getMessage } from '@/app/_utils/message';
// // import { useCaseAzureFunctions } from '../../../../../azure-functions';
// import { crisisManagementScenariosContainer } from '../../../../../cosmos';

// type IdeaResponse = {
//   answer: string;
//   log: LLMserviceBackEndLog<'idea'>;
// };

// type IdeaErrorResponse = {
//   error: string;
// };

// // ダミーデータを生成する関数
// function generateDummyResponse(): IdeaResponse {
//   // テストレスポンス
//   const templateContent = `
// # 危機管理シナリオ

// ## 1. テスト
// - テスト１
// - テスト２
// - テスト３
// `;

//   // ダミーのログデータ
//   const log = {
//     type: 'idea',
//     model: 'dummy-model',
//     prompt: 'これはテスト用のプロンプトです',
//     maxTokens: 1000,
//     result: 'これはテスト用の結果です',
//     totalTokens: 500,
//     startedAt: new Date().toISOString(),
//     completedAt: new Date().toISOString(),
//   } as unknown as LLMserviceBackEndLog<'idea'>;

//   return {
//     answer: templateContent,
//     log,
//   };
// }

// export async function CrisisManagementScenarios(
//   id: string,
//   industry: string,
//   businessSize: string,
//   businessContent: string,
//   selectedOptions: string[],
//   additionalContents?:string,
//   additionalConsiderations?: string
// ): Promise<IdeaResponse | IdeaErrorResponse> {
//   const user = await getCurrentUser();
//   try {
//     console.log('テスト用パラメータ:', {
//       industry,
//       businessSize,
//       businessContent,
//       selectedOptions,
//       additionalContents,
//       additionalConsiderations,
//     });

//     // ダミーデータを生成
//     const response = generateDummyResponse();
//     console.log('ダミーレスポンス生成完了');

//     // レスポンスデータのサイズをチェック
//     try {
//       const responseSize = Buffer.byteLength(JSON.stringify(response), 'utf8');
//       console.log('応答データサイズ:', responseSize, 'bytes');
//     } catch (sizeError) {
//       console.error('サイズ計算エラー:', sizeError);
//     }

//     // 応答を文字列として安全に処理
//     const safeAnswer =
//       typeof response.answer === 'string' ? response.answer : JSON.stringify(response.answer);

//     // log
//     try {
//       await crisisManagementScenariosDB(crisisManagementScenariosContainer, {
//         id,
//         userId: user.id,
//         useName: user.name,
//         userEmail: user.email,
//         userDepartmentName: user.departmentName,
//         createdAt: new Date(),
//         industry,
//         businessSize,
//         businessContent,
//         selectedOptions,
//         additionalContents,
//         additionalConsiderations,
//         outputForm: safeAnswer,
//         log: response.log,
//       });
//     } catch (dbError) {
//       console.error('データベース保存エラー:', dbError);
//       // DBエラーでもフロントエンドには結果を返す
//     }

//     return {
//       answer: safeAnswer,
//       log: response.log,
//     };
//   } catch (error) {
//     console.error('危機管理シナリオ作成error 詳細:', error);

//     if (error instanceof Error) {
//       console.error('エラーメッセージ:', error.message);
//       console.error('エラースタック:', error.stack);

//       // エラーメッセージから有用な情報を抽出
//       if (error.message.includes('exceeded the size limit')) {
//         return { error: 'レスポンスサイズが制限を超えています。より短い入力で試してください。' };
//       }

//       // 正規表現を使ってJSON部分を抽出
//       const jsonRegex = /\{[\s\S]*\}/; // 波括弧から始まり波括弧で終わる部分を抽出
//       const jsonMatch = error.message.match(jsonRegex);

//       if (jsonMatch) {
//         try {
//           const errorObj = JSON.parse(jsonMatch[0]);
//           return { error: errorObj.error_message };
//         } catch (parseError) {
//           console.error('JSONパースエラー:', parseError);
//         }
//       }
//     }
//     return { error: getMessage('E_F_00110', '作成結果') };
//   }
// }
